### ISSUE TITLE
Issue with serialzable objects on mono

### ISSUE THREAD

**Issue by amirtuval** [JamesNK/Newtonsoft.Json#656]
Hi

I ran into an issue with deserializing serializable objects in mono.
I believe the root cause is with different behavior of the SerializationInfo class in micorosft.NET and mono.

I have this class 

```C#
[Serializable]
class Test : ISerializable {

	string m_A;

	public Test(string a) {
		m_A = a;
	}

	protected Test(SerializationInfo info, StreamingContext context) {
		m_A = info.GetString ("a");
	}

	public void GetObjectData (SerializationInfo info, StreamingContext context)
	{
		info.AddValue ("a", m_A);
	}

	public override string ToString() {
		return m_A;
	}
}
```

I try to serialize it to json and then deserialize it.

On microsoft.net it works fine, but on mono I get the following exception

```
System.InvalidCastException: Cannot cast from source type to destination type.
at Newtonsoft.Json.Serialization.JsonFormatterConverter.GetTokenValue<string> (object) 
```

Any help on this would be much appreciated.

Thanks
Amir

**Comment by JamesNK** [comment]
I can't fix mono. Raise the bug with them.

**Comment by amirtuval** [comment]
Well, mono's SerializationInfo class does behave differently, but it is not necessarily a bug.

What happens is that mono's SerializationInfo implementation invokes the converter twice.
If you invoke serializationInfo.GetString, for example, it will first ask to convert JValue to a string.
Afterwards, it will ask to convert the result to a string again, meaning, it will basically ask the converter to convert a string object to a string. Weird behavior, true, but not incorrect.

The bug, I think, is in the JsonFormatterConverter class, which assumes the object given to it is a JValue, without checking.

I forked and made a fix.

If you are interested, you can see it in https://github.com/amirtuval/Newtonsoft.Json.git.
I made the fix off the 6.0.4 tag, since that's the version we are using, so there's a new tag = 6.0.4-with_mono_fix.

I can also send you a pull request, if you are interested.

Thanks
Amir

**Comment by yinso** [comment]
Running into this issue as well, any interest in address it?
